src/dtgtk/utility.h

```diff
@@ -0,0 +1,39 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2020 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <gtk/gtk.h>
+
+// check whether the given container has any user-added children
+gboolean dtgtk_container_has_children(GtkContainer *container);
+// return a count of the user-added children in the given container
+int dtgtk_container_num_children(GtkContainer *container);
+// return the first child of the given container
+GtkWidget *dtgtk_container_first_child(GtkContainer *container);
+// return the requested child of the given container, or NULL if it has fewer children
+GtkWidget *dtgtk_container_nth_child(GtkContainer *container, int which);
+
+// remove all of the children we've added to the container.  Any which no longer have any references will
+// be destroyed.
+void dtgtk_container_remove_children(GtkContainer *container);
+
+// delete all of the children we've added to the container.  Use this function only if you are SURE
+// there are no other references to any of the children (if in doubt, use dtgtk_container_remove_children
+// instead; it's a bit slower but safer).
+void dtgtk_container_destroy_children(GtkContainer *container);
```

Comment:
Hm. Theoretically this is correct, but the cases where we want to empty a container but there still is somewhere a ref out there that is still useful afterwards are going to be really limited and they would require special handling anyway and not benefit from either of these utility functions (or explicitly remove/reparent the one widget they want to keep before destroying the rest).

My gut feeling would be that if someone is emptying a container but isn't sure if they'll leave dangling refs, I'd _want_ them to cause a crash and solve the problem rather than create an unusual situation that nobody has dealt with (and a likely memory leak). Too much "better be safe than sorry" code already that nobody understands what special situations they are trying to deal with and that gets copy pasted everywhere.

But OK, that's just MHO first thing in the morning. Feel free to disagree and ignore!

Reply:
We just had to fix a crash for precisely that reason, where a gtk_container_remove got "optimized" into a gtk_widget_destroy....

Reply:
Ah, I thought that crash was caused by the issue we discussed before, calling widget_destroy with two arguments instead of one. But Lua is keeping refs around. Seems unlikely to me that those would still be useful after they are removed from the container, but I'm not going to dive into that whole subsystem just to figure that out.

@wpferguson without having any idea what I'm talking about (so just dismiss me if this won't work for _whatever_ reason) if lua keeps refs to widgets it puts into a container (i.e. doesn't transfer ownership) would it be possible to do what the docs for `gtk_widget_destroy` suggests, i.e.
_"you should connect to the “destroy” signal if you hold a reference to widget and you wish to remove it when this function is called."_ and then notify (without causing a crash) anyone who still wants to access the widget from lua afterwards that it already has been destroyed and might not contain anything useful anymore. But maybe the previous crashes were caused by lua itself double freeing a destroyed widget during cleanup rather than user code trying to access it?

@ralfbrown if lua _can't_ clean up after a destroy, it may be worth adding that to the comment in `dtgtk_container_destroy_children` as something to look out for.

Reply:
Did the crash have anything to do with resetting the exporter?

Each lua widget has a callback running somewhere, so destroying the widget without removing the callback could have problems.

Reply:
Can someone point me to the original issue so that I can get up to speed please?